Try to clean up clause-meaning. (#3896)
modelica/ModelicaSpecification: chapters/equations.tex

\subsection{If-Equations}\label{if-equations}

The syntax of an \lstinline!if!-equation\index{if@\robustinline{if}!equation}\index{then@\robustinline{then}!if-equation@\robustinline{if}-equation}\index{else@\robustinline{else}!if-equation@\robustinline{if}-equation}\index{elseif@\robustinline{elseif}!if-equation@\robustinline{if}-equation} is given by \productionref{if-equation} in the grammar.

> An \lstinline!if!-statement is seen as a list of branches: one \lstinline!if!-branch, and zero or more \lstinline!elseif!-branches, and an optional \lstinline!else!-branch.

Reviewer: It's an equation:

```suggestion
An \lstinline!if!-equation is seen as a list of branches: one \lstinline!if!-branch, and zero or more \lstinline!elseif!-branches, and an optional \lstinline!else!-branch.
```

> One or zero of the branches is selected, by evaluating their conditions sequentially until a condition that evaluates to true is found.

Reviewer: Align with `if`-statement language:

```suggestion
One or zero of these branches is selected, by evaluating their conditions sequentially until a condition that evaluates to true is found.
```

\subsection{If-Equations}\label{if-equations}

> The syntax of an \lstinline!if!-equations\index{if@\robustinline{if}!equation}\index{then@\robustinline{then}!if-equation@\robustinline{if}-equation}\index{else@\robustinline{else}!if-equation@\robustinline{if}-equation}\index{elseif@\robustinline{elseif}!if-equation@\robustinline{if}-equation} is given by \productionref{if-equation} in the grammar.

Reviewer: Same typo as for the statement:

```suggestion
The syntax of an \lstinline!if!-equation\index{if@\robustinline{if}!equation}\index{then@\robustinline{then}!if-equation@\robustinline{if}-equation}\index{else@\robustinline{else}!if-equation@\robustinline{if}-equation}\index{elseif@\robustinline{elseif}!if-equation@\robustinline{if}-equation} is given by \productionref{if-equation} in the grammar.
```

> Here, each \productionref{expression} in the grammar (representing the condition of any branch of an \lstinline!if!-clause) must be a scalar \lstinline!Boolean! expression.

Reviewer: Similar to suggestion for `if`-statement (sentence will be restored at the end of the paragraph):

```suggestion
```

> An \lstinline!if!-clause is seen as a list of branches: one \lstinline!if!-branch, and zero or more \lstinline!elseif!-branches, and an optional \lstinline!else!-branch.

Reviewer: Similar to the suggestion for `if`-statements:

```suggestion
An \lstinline!if!-statement is seen as a list of branches: one \lstinline!if!-branch, and zero or more \lstinline!elseif!-branches, and an optional \lstinline!else!-branch.
```

One or zero of the branches is selected, by evaluating their conditions sequentially until a condition that evaluates to true is found.
If none of the conditions evaluate to true the \lstinline!else!-branch is selected if it exists, otherwise no branch is selected.
In an equation section, the equations in the body of the selected branch are seen as equations that must be satisfied.

> The bodies of the branches that are not selected have no effect on that model evaluation.

Reviewer: Restore sentence removed above:

```suggestion
The bodies of the branches that are not selected have no effect on that model evaluation.
Each branch condition (represented by \productionref{expression} in the grammar) must be a scalar \lstinline!Boolean! expression.
```

Reviewer: Same here, the conversation including the suggestion to remove the sentence in its old location appears right below on the _Conversation_ page of this PR – it was meant to appear before.

\end{lstlisting}

\begin{nonnormative}

> The equivalence is conceptual since \lstinline!pre($\ldots$)! of a non discrete-time \lstinline!Real! variable or expression can only be used within a \lstinline!when!-clause body.

Reviewer: I think we are missing a hyphen here?

```suggestion
The equivalence is conceptual since \lstinline!pre($\ldots$)! of a non-discrete-time \lstinline!Real! variable or expression can only be used within a \lstinline!when!-clause body.
```

Author: True, and the only one as far as I can see. Not changed in this PR, but still good to improve.

  \lstinline!when!-equations cannot be nested.
\item
  \lstinline!when!-equations can only occur within \lstinline!if!-equations and \lstinline!for!-equations if the controlling expressions are exclusively evaluable expressions.

>   For \lstinline!if!-equations this requires that the condition of the branch containing the \lstinline!when!-equation and the conditions of all preceding branches are evaluable.

Reviewer: The `else`-branch has no condition:

```suggestion
  For \lstinline!if!-equations this requires that the condition (if any) of the branch containing the \lstinline!when!-equation and the conditions of all preceding branches are evaluable.
```

The operator reinitializes \lstinline!x! with \lstinline!expr! at an event instant.
\lstinline!x! is a component-reference (where any subscripts are evaluable) referring to a \lstinline!Real! variable (or an array of \lstinline!Real! variables) that must be selected as a state (resp., states), i.e., \lstinline!reinit! on \lstinline!x! implies \lstinline!stateSelect = StateSelect.always! on \lstinline!x!.
\lstinline!expr! needs to be type-compatible with \lstinline!x!.

> For any given variable (possibly an array variable), \lstinline!reinit! can only be applied (either to an individual variable or to a part of an array variable) in one \lstinline!when!-clause (applying \lstinline!reinit! to a variable in several \lstinline!when!- or \lstinline!elsewhen!-branches of the same \lstinline!when!-equation is allowed).

Reviewer: I guess the idea is to completely eliminate use of the term `when`-equation?

Author: Yes, as much as possible at least. We can still use it say that there when-equations and when-statements, and possibly to define differences between them. Here we had already defined that reinit can only be in when-equations, so I don't see a need to repeat that.

Reviewer: But currently there is a mix of `when`-equation and `when`-clause here, and since `reinit` is only allowed in `when`-equations it is better to stick with the more precise term:
```suggestion
For any given variable (possibly an array variable), \lstinline!reinit! can only be applied (either to an individual variable or to a part of an array variable) in one \lstinline!when!-equation (applying \lstinline!reinit! to a variable in several \lstinline!when!- or \lstinline!elsewhen!-branches of the same \lstinline!when!-equation is allowed).
```

> For any given variable (possibly an array variable), \lstinline!reinit! can only be applied (either to an individual variable or to a part of an array variable) in one \lstinline!when!-equation (applying \lstinline!reinit! to a variable in several \lstinline!when!- or \lstinline!elsewhen!-branches of the same \lstinline!when!-equation is allowed).

Reviewer: There is no need to enumerate all kinds of branches when it is clear from context:

```suggestion
For any given variable (possibly an array variable), \lstinline!reinit! can only be applied (either to an individual variable or to a part of an array variable) in one \lstinline!when!-equation (applying \lstinline!reinit! to a variable in several branches of the same \lstinline!when!-equation is allowed).
```

(This is similar to how we speak of the `if`-equation branches in the next sentence.)

> If there are multiple \lstinline!reinit! for a variable inside one \lstinline!when!-clause body, they must appear in different branches of an \lstinline!if!-equation (in order that at most one \lstinline!reinit! for the variable is active at any event).

Reviewer: More specific:

```suggestion
If there are multiple \lstinline!reinit! for a variable inside one \lstinline!when!-equation body, they must appear in different branches of an \lstinline!if!-equation (in order that at most one \lstinline!reinit! for the variable is active at any event).
```

In case of \lstinline!reinit! active during initialization (due to \lstinline!when initial()!), see \cref{initialization-initial-equation-and-initial-algorithm}.

\lstinline!reinit! does not break the single assignment rule, because \lstinline!reinit(x, expr)! in equations evaluates \lstinline!expr! to a value, then at the end of the current event iteration step it assigns this value to \lstinline!x! (this copying from values to reinitialized state(s) is done after all other evaluations of the model and before copying \lstinline!x! to \lstinline!pre(x)!).

\begin{example}
If a higher index system is present, i.e., constraints between state variables, some state variables need to be redefined to non-state variables.

> During simulation, non-state variables should be chosen in such a way that variables with an applied \lstinline!reinit! are selected as states at least when the corresponding \lstinline!when!-clause bodies become active.

Reviewer: Should it be `when`-clause branches here?

Author: I don't think so. Yes, we _select_ a branch in some sense (based on its condition - and the preceding ones) - but only the body becomes _active_.

Reviewer: Now when I come back after my first sweep, I am more convinced that speaking of activated branches sounds better to me.  One reason I like speaking of _branch_ here is that _branch_ also works when talking about triggering, while _triggering a body_ sounds plain wrong to me.  Also, there is no risk of misunderstanding what activating a branch means.

More generally, I'd also say that it is the branches which contain the bodies, so speaking of a `when`-clause body or `if`-equation body becomes a little bit unclear since it could be mistaken for the union of the bodies of all the branches.

Anyway, an easy way out here is to speak directly of the `reinit`-equation:
```suggestion
During simulation, non-state variables should be chosen in such a way that variables with an applied \lstinline!reinit! are selected as states at least during the events when the \lstinline!reinit!-equation is active.
```

Clocked discrete-time partitions are also treated this way, see \cref{clocked-discrete-time-and-clocked-discretized-continuous-time-partition}.

\begin{nonnormative}

> Expressions in \lstinline!when!-clause bodies do not trigger state events since they are not active during during continuous integration.

Reviewer: Typo:

```suggestion
Expressions in \lstinline!when!-clause bodies do not trigger state events since they are not active during continuous integration.
```

> Using state events in a \lstinline!when!-clause body is unnecessary because the \lstinline!when!-clause body is not evaluated during continuous integration.

Reviewer: Is it even meaningful to speak of using state events in a `when`-clause body?  Wouldn't it be more to the point to explain that expressions in `when`-clause bodies cannot trigger state events?

Author: I would postpone that; as we already had that text.
I believe the intent is to explain _why_ state events are unnecessary here, not explain that Modelica don't have them here; but it could certainly be explained better.

Reviewer: Trying to avoid the need to remember to return to this in a follow-up PR:
```suggestion
Expressions in \lstinline!when!-clause bodies do not trigger state events since they are not active during during continuous integration.
```

  end when;
\end{lstlisting}

> The \lstinline!slowSample! \lstinline!when!-clause body is evaluated at every 5th occurrence of the evaluating the \lstinline!fastSample! \lstinline!when!-clause body.

Reviewer: Something is wrong with the grammar here: _occurrence of the evaluating the_

Should it be simply this?

```suggestion
The \lstinline!slowSample! \lstinline!when!-clause body is evaluated at every 5th evaluation of the \lstinline!fastSample! \lstinline!when!-clause body.
```

Author: Makes sense.

The algorithmic statements within a \lstinline!when!-statement are active during initialization, if and only if they are explicitly enabled with \lstinline!initial()!, and only in one of the two forms \lstinline!when initial() then! or \lstinline!when {$\ldots$, initial(), $\ldots$} then!.
In this case, the algorithmic statements within the \lstinline!when!-statement remain active during the whole initialization phase.

> An active \lstinline!when!-clause branch condition inactivates the following \lstinline!elsewhen! (similarly to \lstinline!when!-clauses during simulation), but apart from that the first \lstinline!elsewhen initial() then! or \lstinline!elsewhen {$\ldots$, initial(), $\ldots$} then! is similarly active during initialization as \lstinline!when initial() then! or \lstinline!when {$\ldots$, initial(), $\ldots$} then!.

Reviewer: Write out _branches_:

```suggestion
An active \lstinline!when!-clause branch condition inactivates the following \lstinline!elsewhen!-branches (similarly to \lstinline!when!-clauses during simulation), but apart from that the first \lstinline!elsewhen initial() then! or \lstinline!elsewhen {$\ldots$, initial(), $\ldots$} then! is similarly active during initialization as \lstinline!when initial() then! or \lstinline!when {$\ldots$, initial(), $\ldots$} then!.
```

> An active condition of a branch of a \lstinline!when!-clause inactivates the following \lstinline!elsewhen! (similarly to \lstinline!when!-clauses during simulation), but apart from that the first \lstinline!elsewhen initial() then! or \lstinline!elsewhen {$\ldots$, initial(), $\ldots$} then! is similarly active during initialization as \lstinline!when initial() then! or \lstinline!when {$\ldots$, initial(), $\ldots$} then!.

Reviewer: Simplify language:

```suggestion
An active \lstinline!when!-clause branch condition inactivates the following \lstinline!elsewhen! (similarly to \lstinline!when!-clauses during simulation), but apart from that the first \lstinline!elsewhen initial() then! or \lstinline!elsewhen {$\ldots$, initial(), $\ldots$} then! is similarly active during initialization as \lstinline!when initial() then! or \lstinline!when {$\ldots$, initial(), $\ldots$} then!.
```

\end{lstlisting}
After solving the initialization problem, both \lstinline!b! and \lstinline!pre(b)! are false.

> During transient analysis, \lstinline!initial()! is false, so \lstinline!b! changes from falso to true during the first round of the initial event iteration.

Reviewer: Nearby typo:

```suggestion
During transient analysis, \lstinline!initial()! is false, so \lstinline!b! changes from false to true during the first round of the initial event iteration.
```

> This triggers the \lstinline!when!-clause body, and \lstinline!x! gets reinitialized.

Reviewer: Isn't it better to speak of triggered branches than triggered bodies?

```suggestion
This triggers the \lstinline!when!-clause branch, and \lstinline!x! gets reinitialized.
```

Author: As above I think that only the body is active/triggered.

Reviewer: I'd say it's especially when speaking of triggering that _branch_ makes much better sense than _body_.

Author: Thinking more I believe that a solution might be  something _subtly_ different (which indicates that more time is needed). 

I still think that only the body is really triggered, whereas the condition is triggering it. 

Both of these are part of the branch, so to me it seems that the branch should be more active (instead of passively getting triggered). Perhaps something like:
```suggestion
The \lstinline!when!-clause branch triggers, and \lstinline!x! gets reinitialized.
```
possibly replace "triggers" by "triggers based on the change in b".

Reviewer: The difference is very subtle and I don't see why the active way to see it would make _more_ sense than the passive, while at least the passive is more easy to concisely connect with the previous sentence.

Author: ```suggestion
This activates the \lstinline!when!-clause body, and \lstinline!x! gets reinitialized.
```
